connect-location/src/main/java/com/ifttt/location/AwarenessGeofenceProvider.java

```diff
@@ -56,26 +58,21 @@
     public void updateGeofences(final List<Feature> features) {
         final List<String> updatedIds = new ArrayList<>();
 
-        List<UserFeatureStep> userFeatureSteps = new ArrayList<>();
+        List<UserFeatureField> userFeatureFields = new ArrayList<>();
 
         for (Feature feature : features) {
-            if (feature.userFeatures != null) {
-                for (UserFeature userFeature : feature.userFeatures) {
-                    if (userFeature.userFeatureSteps != null) {
-                        userFeatureSteps.addAll(userFeature.userFeatureSteps);
-                    }
-                }
+            if (feature.userFeatures == null) {
+                continue;
             }
-        }
 
-        List<UserFeatureField> userFeatureFields = new ArrayList<>();
+            for (UserFeature userFeature : feature.userFeatures) {
```

Comment:
If a UserFeature is non-null, it should have a non-null steps list, each step should have a non-null list of fields, and each field should have a non-null value. 

If any of these is not true, the connection is going to be in an invalid state.

Reply:
would it make sense to throw exceptions for these invalid states?

Reply:
The exceptions are thrown already in the JSON adapter. I don't know if we want to crash the host app here though, since it's most likely not caused by the host app. 

Currently the way we handle this is to throw a generic error in Retrofit level. I see that we can improve it for debugging purpose, but will be out of the scope for this PR.